bot/code_review_bot/revisions.py

```diff
@@ -180,6 +184,67 @@ def from_try(try_task: dict, phabricator: PhabricatorAPI):
             patch=patch,
         )
 
+    @staticmethod
+    def from_autoland(autoland_task: dict, phabricator: PhabricatorAPI):
+        """
+        Build a revision from a Mozilla autoland decision task
+        """
+        assert (
+            autoland_task["payload"]["env"]["GECKO_HEAD_REPOSITORY"] == REPO_AUTOLAND
+        ), "Not an autoland decision task"
+
+        # Load mercurial revision
+        mercurial_revision = autoland_task["payload"]["env"]["GECKO_HEAD_REV"]
+
+        # Search phabricator revision from commit message
+        commit_url = (
+            f"https://hg.mozilla.org/integration/autoland/json-rev/{mercurial_revision}"
+        )
+        response = requests.get(commit_url)
+        response.raise_for_status()
+        description = response.json()["desc"]
+        match = REGEX_PHABRICATOR_COMMIT.search(description)
+        if match is not None:
+            url, revision_id = match.groups()
+            revision_id = int(revision_id)
+            logger.info("Found phabricator revision", id=revision_id, url=url)
+        else:
+            raise Exception(f"No phabricator revision found in commit {commit_url}")
+
+        # Lookup the Phabricator revision to get details (phid, title, bugzilla_id, ...)
+        revision = phabricator.load_revision(rev_id=revision_id)
+
+        # Search the Phabricator diff with same commit identifier
+        diffs = phabricator.search_diffs(
+            revision_phid=revision["phid"], attachments={"commits": True}
+        )
+        diff = next(
+            iter(
+                d
+                for d in diffs
+                if d["attachments"]["commits"]["commits"][0]["identifier"]
+                == mercurial_revision
+            ),
+            None,
+        )
+        assert (
+            diff is not None
+        ), f"No Phabricator diff found for D{revision_id} and mercurial revision {mercurial_revision}"
+        logger.info("Found phabricator diff", id=diff["id"])
```

Comment:
Why do we need the Phabricator revision/diff in the autoland case? Is it just to avoid larger changes in the bot?

Reply:
We need it to create the `Diff` on the backend: it's the `Diff` that stores the mercurial revision + repository (hence indicating the Phabricator revision has reached autoland)